### PR TITLE
Popup and Footnotes docs update and small bugfix

### PR DESCRIPTION
**Pull request by dshovchko**

### CODE REVIEW

File: src/modules/esl-tooltip/core/esl-tooltip.ts [hunk 1]
```diff
@@ -62,21 +76,25 @@ export class ESLTooltip extends ESLPopup {
...
   @memoize()
-  protected _createArrow() {
+  protected _createArrow(): HTMLElement {
     const arrow = document.createElement('span');
     arrow.className = 'esl-popup-arrow';
     return arrow;
   }
 
-  protected _appendArrow() {
+  /** Appent arrow to Tooltip */
```
Comment (ala-n) [inline src/modules/esl-tooltip/core/esl-tooltip.ts:90]:
```suggestion
  /** Appends arrow to Tooltip */
```

File: src/modules/esl-tooltip/core/esl-tooltip.ts [hunk 2]
```diff
@@ -62,21 +76,25 @@ export class ESLTooltip extends ESLPopup {
     this.tabIndex = 0;
   }
 
+  /** Set initial state of the Tooltip */
   protected setInitialState() {}
 
+  /** Create arrow at Tooltip */
```
Comment (ala-n) [inline src/modules/esl-tooltip/core/esl-tooltip.ts:82]:
```suggestion
  /** Creates arrow at Tooltip */
```

File: src/modules/esl-tooltip/core/esl-tooltip.ts [hunk 3]
```diff
@@ -23,37 +23,51 @@ export interface TooltipActionParams extends PopupActionParams {
...
   }
 
-  public static show(params: TooltipActionParams = {}) {
+  /** Change the element state to active */
+  public static show(params: TooltipActionParams = {}): void {
     ESLTooltip.sharedInstance.show(params);
   }
 
-  public static hide(params: TooltipActionParams = {}) {
+  /** Change the element state to inactive */
```
Comment (ala-n) [inline src/modules/esl-tooltip/core/esl-tooltip.ts:65]:
```suggestion
  /** Changes the element state to inactive */
```

File: src/modules/esl-tooltip/core/esl-tooltip.ts [hunk 4]
```diff
@@ -23,37 +23,51 @@ export interface TooltipActionParams extends PopupActionParams {
...
     return els.length ? els[els.length - 1] : null;
   }
 
+  /** Active state marker */
   public static get open(): boolean {
     return ESLTooltip.sharedInstance.open;
   }
 
-  public static show(params: TooltipActionParams = {}) {
+  /** Change the element state to active */
```
Comment (ala-n) [inline src/modules/esl-tooltip/core/esl-tooltip.ts:60]:
```suggestion
  /** Changes the element state to active */
```

File: src/modules/esl-footnotes/core/esl-note.ts [hunk 5]
```diff
@@ -148,19 +155,19 @@ export class ESLNote extends ESLBaseElement {
     this.highlight();
   }
   /** Hide tooltip with passed params */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-note.ts:157]:
```suggestion
  /** Hides tooltip with passed params */
```

File: src/modules/esl-footnotes/core/esl-note.ts [hunk 6]
```diff
@@ -82,54 +84,59 @@ export class ESLNote extends ESLBaseElement {
...
 
-  public unlink() {
+  /** Unlink note from footnotes */
+  public unlink(): void {
     this.restore();
     this._sendResponseToFootnote();
   }
 
-  protected restore() {
+  /** Restore original note content after unlinking */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-note.ts:130]:
```suggestion
  /** Restores original note content after unlinking */
```

File: src/modules/esl-footnotes/core/esl-note.ts [hunk 7]
```diff
@@ -82,54 +84,59 @@ export class ESLNote extends ESLBaseElement {
...
   }
 
-  public highlight(enable: boolean = true) {
+  /** Highlight note */
+  public highlight(enable: boolean = true): void {
     this.classList.toggle('highlight', enable);
   }
 
-  public link(footnotes: ESLFootnotes, index: number) {
+  /** Link note with footnotes */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-note.ts:116]:
```suggestion
  /** Links note with footnotes */
```

File: src/modules/esl-footnotes/core/esl-note.ts [hunk 8]
```diff
@@ -82,54 +84,59 @@ export class ESLNote extends ESLBaseElement {
...
+  protected unbindEvents(): void {
     document.body.removeEventListener(`${ESLFootnotes.eventNs}:request`, this._onFootnotesReady);
     this.removeEventListener('click', this._onClick);
     this.removeEventListener('keydown', this._onKeydown);
     this.removeEventListener('mouseenter', this._onMouseEnter);
     this.removeEventListener('mouseleave', this._onMouseLeave);
   }
 
-  public activate() {
+  /** Activate note */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-note.ts:102]:
```suggestion
  /** Activates note */
```

File: src/modules/esl-footnotes/core/esl-note.ts [hunk 9]
```diff
@@ -46,6 +46,7 @@ export class ESLNote extends ESLBaseElement {
     return ['tooltip-shown'];
   }
 
+  /** Note index */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-note.ts:49]:
```suggestion
  /** Note index in the scope content */
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 10]
```diff
@@ -146,13 +161,17 @@ export class ESLFootnotes extends ESLBaseElement {
...
     this._notes
       .filter((item) => note.html === item.html)
       .forEach((item) => item.highlight(false));
   }
 
-  protected _sendRequestToNote() {
+  /**
+   * Send a request to all notes, expecting to get a response from
+   * the unlinked ones and link up with them */
+  protected _sendRequestToNote(): void {
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:174]: Just an idea for naming: `_notifyNotes()`

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 11]
```diff
@@ -146,13 +161,17 @@ export class ESLFootnotes extends ESLBaseElement {
...
+  /** Turn off highlight from notes with the same text */
+  public turnOffHighlight(note: ESLNote): void {
     this._notes
       .filter((item) => note.html === item.html)
       .forEach((item) => item.highlight(false));
   }
 
-  protected _sendRequestToNote() {
+  /**
+   * Send a request to all notes, expecting to get a response from
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:172]:
```suggestion
   * Sends a request to all notes, expecting to get a response from
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 12]
```diff
@@ -59,70 +62,80 @@ export class ESLFootnotes extends ESLBaseElement {
...
   protected buildItemIndex(footnote: FootnotesItem): string {
     return `<span class="esl-footnotes-index">${footnote.renderedIndex.join(', ')}</span>`;
   }
 
+  /** Build item text */
   protected buildItemText(footnote: FootnotesItem): string {
     return `<span class="esl-footnotes-text">${footnote.text}</span>`;
   }
 
+  /** Build item back-to-note button */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:121]:
```suggestion
  /** Builds item back-to-note button */
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 13]
```diff
@@ -59,70 +62,80 @@ export class ESLFootnotes extends ESLBaseElement {
...
     const item = `${this.buildItemIndex(footnote)}${this.buildItemText(footnote)}${this.buildItemBack(footnote)}`;
     return `<li class="esl-footnotes-item" data-order="${footnote.index}">${item}</li>`;
   }
 
+  /** Build item index */
   protected buildItemIndex(footnote: FootnotesItem): string {
     return `<span class="esl-footnotes-index">${footnote.renderedIndex.join(', ')}</span>`;
   }
 
+  /** Build item text */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:116]:
```suggestion
  /** Builds item text */
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 14]
```diff
@@ -59,70 +62,80 @@ export class ESLFootnotes extends ESLBaseElement {
...
     return `<ul class="esl-footnotes-items">${items}</ul>`;
   }
 
+  /** Build one item from footnotes list */
   protected buildItem(footnote: FootnotesItem): string {
     const item = `${this.buildItemIndex(footnote)}${this.buildItemText(footnote)}${this.buildItemBack(footnote)}`;
     return `<li class="esl-footnotes-item" data-order="${footnote.index}">${item}</li>`;
   }
 
+  /** Build item index */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:111]:
```suggestion
  /** Builds item index */
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 15]
```diff
@@ -59,70 +62,80 @@ export class ESLFootnotes extends ESLBaseElement {
...
     this.innerHTML = this.buildItems();
   }
 
+  /** Build content of footnotes */
   protected buildItems(): string {
     const items = this.footnotesList.map((footnote) => this.buildItem(footnote)).join('');
     return `<ul class="esl-footnotes-items">${items}</ul>`;
   }
 
+  /** Build one item from footnotes list */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:105]:
```suggestion
  /** Builds one item from footnotes list */
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 16]
```diff
@@ -59,70 +62,80 @@ export class ESLFootnotes extends ESLBaseElement {
...
     this.update();
   }
 
-  public update() {
+  /** Update content of footnotes */
+  public update(): void {
     this.innerHTML = this.buildItems();
   }
 
+  /** Build content of footnotes */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:99]:
```suggestion
  /** Builds content of footnotes */
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 17]
```diff
@@ -59,70 +62,80 @@ export class ESLFootnotes extends ESLBaseElement {
...
 
-  public unlinkNote(note: ESLNote) {
+  /** Remove note from footnotes list */
+  public unlinkNote(note: ESLNote): void {
     this._notes = this._notes.filter((el) => el !== note);
     this.update();
   }
 
-  public update() {
+  /** Update content of footnotes */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:94]:
```suggestion
  /** Updates the content of footnotes */
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 18]
```diff
@@ -59,70 +62,80 @@ export class ESLFootnotes extends ESLBaseElement {
...
+  /** Add note to footnotes list */
+  public linkNote(note: ESLNote): void {
     if (this._notes.includes(note)) return;
     this._notes.push(note);
     const index = +sequentialUID(ESLFootnotes.is, '');
     note.link(this, index);
   }
 
-  public unlinkNote(note: ESLNote) {
+  /** Remove note from footnotes list */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:88]:
```suggestion
  /** Removes the note from the footnotes list */
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 19]
```diff
@@ -59,70 +62,80 @@ export class ESLFootnotes extends ESLBaseElement {
...
+  protected unbindEvents(): void {
     if (this.scopeEl) {
       this.scopeEl.removeEventListener(`${ESLFootnotes.eventNs}:response`, this._onNoteSubscribe);
     }
     this.removeEventListener('click', this._onClick);
     this.removeEventListener('keydown', this._onKeydown);
   }
 
-  public linkNote(note: ESLNote) {
+  /** Add note to footnotes list */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:80]:
```suggestion
  /** Adds the note to the footnotes list */
```

File: src/modules/esl-footnotes/core/esl-footnotes.ts [hunk 20]
```diff
@@ -27,19 +27,22 @@ export class ESLFootnotes extends ESLBaseElement {
...
+  /** List of notes to show */
   protected get footnotesList(): FootnotesItem[] {
     this.reindex();
     return this.grouping !== 'enable'
       ? compileFootnotesNongroupedList(this._notes)
       : compileFootnotesGroupedList(this._notes);
   }
 
-  public reindex() {
+  /** Reindex list of notes */
```
Comment (ala-n) [inline src/modules/esl-footnotes/core/esl-footnotes.ts:44]:
```suggestion
  /** Reindexes the list of notes */
```

**Review by abarmina — APPROVED** [review]

**Review by NastaLeo — APPROVED** [review]

File: src/modules/esl-footnotes/core/esl-footnotes.ts (hunk 10 above)
Comment (dshovchko) [reply src/modules/esl-footnotes/core/esl-footnotes.ts:174]: That's a great proposition.

**Review by ala-n — APPROVED** [review]

**Pull request merged**